### ISSUE TITLE
Add new no-self-assign rule

### ISSUE THREAD

**Issue by othomann** [eslint/eslint#4729]
**When does this rule warn? Please describe and show example code:**
This rule would warn when assigning a variable to itself.
var x = 10; x = x;
or
var x = x;
It can be useful when you have a constant value that has the same name as a variable but not the same case and by mistake you assign the variable to itself instead of using the constant.

**Is this rule preventing an error or is it stylistic?**
It is very similar to the no-self-compare existing rule.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**
Since it is really close to the no-self-compare rule, I think it has the same intend.

**Are you willing to create the rule yourself?**
Sure I could contribute the rule myself.

**Comment by eslintbot** [comment]
@othomann Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
I thought we already had a rule that did this. @mysticatea ?

**Comment by ilyavolodin** [comment]
We have `no-self-compare`. We also have a rule that will flag `var x = x;` (I think it's `no-undef`), but I don't think we have anything to cover `var x = 10; x = x;` situation. I'm not sure it's necessary, though, since it's not harmful in any way and too much of an edge case, in my opinion.

**Comment by Slayer95** [comment]
This tends to happen after refactoring code. Automatically catching it with ESLint would great.

**Comment by mysticatea** [comment]
`var x = x;` pattern is caught by `no-use-before-define` since `v2.0.0-alpha-1`. Tests is here: https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-use-before-define.js#L81-L90

`x = x` pattern is not caught now.
But, there are well-known patterns (e.g. `foo = foo || {};`), so I'm not sure whether or not the rule which catches the pattern is proper.

**Comment by nzakas** [comment]
I think the intent of this proposal is to only warn when the assignment is exact (`x = x`).

**Comment by mysticatea** [comment]
OK, I think it's reasonable.

How about destructuring?

```js
[a, b] = [a, b]

// also,
[a, b] = [a, c]
```

**Comment by nzakas** [comment]
It would be good to cover that case, too.

**Comment by mysticatea** [comment]
I'm working on this.
